requestSeqNo does not match the received sequence number error

We are frequently hitting the below error when reading from the eventhub. We see this on both DataBricks and HDInsights.
Version: azure-eventhubs-spark_2.11-2.3.2

Callstack:

WARN CachedEventHubsReceiver: 51691860 did not match 51690986.Recreating receiver for {"ehName":"rawvsevents","partitionId":2}
...

ERROR Executor: Exception in task 2.1 in stage 81.0 (TID 24586)
java.lang.AssertionError: assertion failed: requestSeqNo 51691860 does not match the received sequence number 51690986
	at scala.Predef$.assert(Predef.scala:170)
	at org.apache.spark.eventhubs.client.CachedEventHubsReceiver.org$apache$spark$eventhubs$client$CachedEventHubsReceiver$$receive(CachedEventHubsReceiver.scala:123)
	at org.apache.spark.eventhubs.client.CachedEventHubsReceiver$.receive(CachedEventHubsReceiver.scala:157)
	at org.apache.spark.eventhubs.rdd.EventHubsRDD$EventHubsRDDIterator.next(EventHubsRDD.scala:147)
	at org.apache.spark.eventhubs.rdd.EventHubsRDD$EventHubsRDDIterator.next(EventHubsRDD.scala:128)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:409)
	at org.apache.spark.util.collection.ExternalSorter.insertAll(ExternalSorter.scala:193)
	at org.apache.spark.shuffle.sort.SortShuffleWriter.write(SortShuffleWriter.scala:63)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:96)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.scheduler.Task.run(Task.scala:109)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

--- a/core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala
+++ b/core/src/main/scala/org/apache/spark/eventhubs/client/CachedEventHubsReceiver.scala
@@ -74,7 +74,9 @@ private[client] class CachedEventHubsReceiver private (ehConf: EventHubsConf,
   private def closeReceiver(): Unit = {
     Try(_receiver.closeSync()) match {
       case Success(_) => _receiver = null
-      case Failure(e) => logInfo("closeSync failed in cached receiver.", e)
+      case Failure(e) =>
+        logInfo("closeSync failed in cached receiver.", e)
+        _receiver = null
     }
   }
 
@@ -90,30 +92,33 @@ private[client] class CachedEventHubsReceiver private (ehConf: EventHubsConf,
     s"requestSeqNo $requestSeqNo does not match the received sequence number $receivedSeqNo"
 
   private def receive(requestSeqNo: SequenceNumber, batchSize: Int): EventData = {
-    var event: EventData = null
-    var i: java.lang.Iterable[EventData] = null
-    while (i == null) {
-      i = try {
-        receiver.receiveSync(1)
-      } catch {
-        case r: ReceiverDisconnectedException =>
-          throw new Exception(
-            "You are likely running multiple Spark jobs with the same consumer group. " +
-              "For each Spark job, please create and use a unique consumer group to avoid this issue.",
-            r
-          )
+    def receiveOneEvent: EventData = {
+      var event: EventData = null
+      var i: java.lang.Iterable[EventData] = null
+      while (i == null) {
+        i = try {
+          receiver.receiveSync(1)
+        } catch {
+          case r: ReceiverDisconnectedException =>
+            throw new Exception(
+              "You are likely running multiple Spark jobs with the same consumer group. " +
+                "For each Spark job, please create and use a unique consumer group to avoid this issue.",
+              r
+            )
+        }
       }
+      event = i.iterator.next
+      event
     }
-    event = i.iterator.next
 
+    var event: EventData = receiveOneEvent
     if (requestSeqNo != event.getSystemProperties.getSequenceNumber) {
       logWarning(
         s"$requestSeqNo did not match ${event.getSystemProperties.getSequenceNumber}." +
           s"Recreating receiver for $nAndP")
       closeReceiver()
       createReceiver(requestSeqNo)
-      while (i == null) { i = receiver.receiveSync(1) }
-      event = i.iterator.next
+      event = receiveOneEvent
       assert(requestSeqNo == event.getSystemProperties.getSequenceNumber,
              errWrongSeqNo(requestSeqNo, event.getSystemProperties.getSequenceNumber))
     }

